# Project 3: Semi-Structure Data Analytics
                        MongoDB sample: sample_supplies (5 Queries)

### I. Ingestion

How the data is brought in using json as the data source:

On MongoDB cluster 0 – retrieve connection string 

+ (Add mongo DB connection)  


In Terminal:

- cd C:\Users\User\Desktop\AU Junior\Big Data\Project 3 

- Mongo Export: 
mongoexport --uri="mongodb+srv://u6411242:bdaclass10pass@cluster0.0xdu10r.mongodb.net/sample_supplies" --collection="sales" --out=sample_supplies.json 


- Class 12: mongoexport --uri="mongodb+srv://test-airbnb:test-airbnb@cluster0.9jsifuz.mongodb.net/sample_mflix" --collection="movies" --out=movies.json 

### II. Transformation

In [1]:
import pandas as pd

# Import the data from the JSON file into a Pandas DataFrame
salesdf = pd.read_json('sample_supplies.json', lines=True)

Exploring the data, About Nesting (JSON) & Denesting

In [2]:
salesdf.head()

_id  \
0  {'$oid': '5bd761dcae323e45a93ccfe8'}   
1  {'$oid': '5bd761dcae323e45a93ccfe9'}   
2  {'$oid': '5bd761dcae323e45a93ccfea'}   
3  {'$oid': '5bd761dcae323e45a93ccfeb'}   
4  {'$oid': '5bd761dcae323e45a93ccfec'}   

                                saleDate  \
0  {'$date': '2015-03-23T21:06:49.506Z'}   
1  {'$date': '2015-08-25T10:01:02.918Z'}   
2  {'$date': '2017-06-22T09:54:14.185Z'}   
3  {'$date': '2015-02-23T09:53:59.343Z'}   
4  {'$date': '2017-12-03T18:39:48.253Z'}   

                                               items storeLocation  \
0  [{'name': 'printer paper', 'tags': ['office', ...        Denver   
1  [{'name': 'envelopes', 'tags': ['stationary', ...       Seattle   
2  [{'name': 'notepad', 'tags': ['office', 'writi...        Denver   
3  [{'name': 'binder', 'tags': ['school', 'genera...       Seattle   
4  [{'name': 'backpack', 'tags': ['school', 'trav...        London   

                                            customer  couponUsed  \
0  {'gender': 'M', 'age': 42, 'email': 'cauho@wit...        True   
1  {'gender': 'M', 'age': 50, 'email': 'keecade@h...       False   
2  {'gender': 'M', 'age': 51, 'email': 'worbiduh@...       False   
3  {'gender': 'F', 'age': 45, 'email': 'vatires@t...       False   
4  {'gender': 'M', 'age': 40, 'email': 'dotzu@ib....       False   

  purchaseMethod  
0         Online  
1          Phone  
2       In store  
3       In store  
4       In store

One thing to note is the JSON structure of Nesting

EX: Columns where there are bracket [], its holding arrays or lists in JSON

In [3]:
# To fix this:
# We "Flatten" [] brackets instead of listing multiple products becomes its own row of info
salesdf = salesdf.explode('items')

In [4]:
#Try again to see our changes
salesdf.head()

_id  \
0  {'$oid': '5bd761dcae323e45a93ccfe8'}   
0  {'$oid': '5bd761dcae323e45a93ccfe8'}   
0  {'$oid': '5bd761dcae323e45a93ccfe8'}   
0  {'$oid': '5bd761dcae323e45a93ccfe8'}   
0  {'$oid': '5bd761dcae323e45a93ccfe8'}   

                                saleDate  \
0  {'$date': '2015-03-23T21:06:49.506Z'}   
0  {'$date': '2015-03-23T21:06:49.506Z'}   
0  {'$date': '2015-03-23T21:06:49.506Z'}   
0  {'$date': '2015-03-23T21:06:49.506Z'}   
0  {'$date': '2015-03-23T21:06:49.506Z'}   

                                               items storeLocation  \
0  {'name': 'printer paper', 'tags': ['office', '...        Denver   
0  {'name': 'notepad', 'tags': ['office', 'writin...        Denver   
0  {'name': 'pens', 'tags': ['writing', 'office',...        Denver   
0  {'name': 'backpack', 'tags': ['school', 'trave...        Denver   
0  {'name': 'notepad', 'tags': ['office', 'writin...        Denver   

                                            customer  couponUsed  \
0  {'gender': 'M', 'age': 42, 'email': 'cauho@wit...        True   
0  {'gender': 'M', 'age': 42, 'email': 'cauho@wit...        True   
0  {'gender': 'M', 'age': 42, 'email': 'cauho@wit...        True   
0  {'gender': 'M', 'age': 42, 'email': 'cauho@wit...        True   
0  {'gender': 'M', 'age': 42, 'email': 'cauho@wit...        True   

  purchaseMethod  
0         Online  
0         Online  
0         Online  
0         Online  
0         Online

(id & storeLocation as reference) shows that it has been split correctly

same id for the first few rows, same location: Denver

### III. Querying the Data

#### Query 1: Top 10 products name based on sales 

Make new column for product names

In [12]:
# Extract product name from 'items' as its nested in a dictionary
# We use item['name'] to access and return the value associated with the 'name' key within the dictionary

def extract_product_name(item):
    return item['name']

salesdf['product_name'] = salesdf['items'].apply(extract_product_name)

Make new column for Sales (quantity x price)

In [17]:
# Extract the 'quantity' and 'price' from the 'items' dictionaries, which are nested within each row. 
# Access the 'quantity' and 'price' value within these dictionaries as follows:
# Calculate the sales (quantity x price) for each product

def calculate_sales(item):
    quantity = item['quantity']
    price = float(item['price']['$numberDecimal'])
    return quantity * price


salesdf['sales'] = salesdf['items'].apply(calculate_sales)

In [18]:
print(salesdf.columns)

Index(['_id', 'saleDate', 'items', 'storeLocation', 'customer', 'couponUsed',
       'purchaseMethod', 'product_name', 'sales'],
      dtype='object')


Now we find Top Ten Products by their Sales

In [19]:
# Group by product name and sum the sales to find the top 10 by sales
top_products_by_sales = salesdf.groupby('product_name')['sales'].sum().nlargest(10).reset_index()

# Rename the columns for clarity
top_products_by_sales.columns = ['Product Name', 'Total Sales']

print(top_products_by_sales)

    Product Name  Total Sales
0         laptop   6775977.07
1       backpack    817374.10
2           pens    581843.27
3         binder    511644.57
4        notepad    463615.48
5      envelopes    376658.49
6  printer paper    367459.29


#### Query 2: Top 3 products sales by store (location)

In [20]:
# Group by 'storeLocation' and 'product_name', summing the sales for each product in each location
top_products_by_store = salesdf.groupby(['storeLocation', 'product_name'])['sales'].sum().reset_index()

# Sort the values to take the top 3 for each store
top_products_by_store = top_products_by_store.groupby('storeLocation').apply(lambda x: x.nlargest(3, 'sales')).reset_index(drop=True)

# Rename the columns
top_products_by_store.columns = ['Store Location', 'Product Name', 'Total Sales']

print(top_products_by_store)

   Store Location Product Name  Total Sales
0          Austin       laptop   1018494.05
1          Austin     backpack    115217.59
2          Austin         pens     74622.52
3          Denver       laptop   1961659.25
4          Denver     backpack    245679.34
5          Denver         pens    185008.39
6          London       laptop   1073098.72
7          London     backpack    144529.14
8          London         pens     92602.30
9        New York       laptop    694613.32
10       New York     backpack     82563.14
11       New York         pens     64029.90
12      San Diego       laptop    462973.58
13      San Diego     backpack     50997.42
14      San Diego         pens     42134.55
15        Seattle       laptop   1565138.15
16        Seattle     backpack    178387.47
17        Seattle         pens    123445.61


So lets breakdown this result: There are 6 locations in total, and when we display 
the top 3 products of each store location, they are the same 3 products: Laptop, Backpack, & Pens

In [21]:
salesdf.head()

_id  \
0  {'$oid': '5bd761dcae323e45a93ccfe8'}   
0  {'$oid': '5bd761dcae323e45a93ccfe8'}   
0  {'$oid': '5bd761dcae323e45a93ccfe8'}   
0  {'$oid': '5bd761dcae323e45a93ccfe8'}   
0  {'$oid': '5bd761dcae323e45a93ccfe8'}   

                                saleDate  \
0  {'$date': '2015-03-23T21:06:49.506Z'}   
0  {'$date': '2015-03-23T21:06:49.506Z'}   
0  {'$date': '2015-03-23T21:06:49.506Z'}   
0  {'$date': '2015-03-23T21:06:49.506Z'}   
0  {'$date': '2015-03-23T21:06:49.506Z'}   

                                               items storeLocation  \
0  {'name': 'printer paper', 'tags': ['office', '...        Denver   
0  {'name': 'notepad', 'tags': ['office', 'writin...        Denver   
0  {'name': 'pens', 'tags': ['writing', 'office',...        Denver   
0  {'name': 'backpack', 'tags': ['school', 'trave...        Denver   
0  {'name': 'notepad', 'tags': ['office', 'writin...        Denver   

                                            customer  couponUsed  \
0  {'gender': 'M', 'age': 42, 'email': 'cauho@wit...        True   
0  {'gender': 'M', 'age': 42, 'email': 'cauho@wit...        True   
0  {'gender': 'M', 'age': 42, 'email': 'cauho@wit...        True   
0  {'gender': 'M', 'age': 42, 'email': 'cauho@wit...        True   
0  {'gender': 'M', 'age': 42, 'email': 'cauho@wit...        True   

  purchaseMethod   product_name   sales  
0         Online  printer paper   80.02  
0         Online        notepad   70.58  
0         Online           pens  280.60  
0         Online       backpack  155.42  
0         Online        notepad   36.94

#### Query 3: Rankings of each store (location)

In [23]:
# Calculate the total sales for each store
store_sales = salesdf.groupby('storeLocation')['sales'].sum().reset_index()

# Rank stores by total sales and create a new column for rankings
store_sales['rank'] = store_sales['sales'].rank(ascending=False, method='min')

# Sort the store rankings
store_sales = store_sales.sort_values(by='rank')

# Reset the index for clarity
store_sales = store_sales.reset_index(drop=True)

print(store_sales)

  storeLocation       sales  rank
0        Denver  2921009.92   1.0
1       Seattle  2255947.69   2.0
2        London  1583066.79   3.0
3        Austin  1445603.11   4.0
4      New York  1016059.59   5.0
5     San Diego   672885.17   6.0


#### Query 4: Show purchased method by gender table

Make a new column to extract gender

In [26]:
# 'gender' from the 'cusomter' dictionaries nested within each row.

def extract_customer_gender(customer):
    return customer.get('gender')

salesdf['gender'] = salesdf['customer'].apply(extract_customer_gender)

In [35]:
salesdf.head()

_id  \
0  {'$oid': '5bd761dcae323e45a93ccfe8'}   
0  {'$oid': '5bd761dcae323e45a93ccfe8'}   
0  {'$oid': '5bd761dcae323e45a93ccfe8'}   
0  {'$oid': '5bd761dcae323e45a93ccfe8'}   
0  {'$oid': '5bd761dcae323e45a93ccfe8'}   

                                saleDate  \
0  {'$date': '2015-03-23T21:06:49.506Z'}   
0  {'$date': '2015-03-23T21:06:49.506Z'}   
0  {'$date': '2015-03-23T21:06:49.506Z'}   
0  {'$date': '2015-03-23T21:06:49.506Z'}   
0  {'$date': '2015-03-23T21:06:49.506Z'}   

                                               items storeLocation  \
0  {'name': 'printer paper', 'tags': ['office', '...        Denver   
0  {'name': 'notepad', 'tags': ['office', 'writin...        Denver   
0  {'name': 'pens', 'tags': ['writing', 'office',...        Denver   
0  {'name': 'backpack', 'tags': ['school', 'trave...        Denver   
0  {'name': 'notepad', 'tags': ['office', 'writin...        Denver   

                                            customer  couponUsed  \
0  {'gender': 'M', 'age': 42, 'email': 'cauho@wit...        True   
0  {'gender': 'M', 'age': 42, 'email': 'cauho@wit...        True   
0  {'gender': 'M', 'age': 42, 'email': 'cauho@wit...        True   
0  {'gender': 'M', 'age': 42, 'email': 'cauho@wit...        True   
0  {'gender': 'M', 'age': 42, 'email': 'cauho@wit...        True   

  purchaseMethod   product_name   sales gender  
0         Online  printer paper   80.02      M  
0         Online        notepad   70.58      M  
0         Online           pens  280.60      M  
0         Online       backpack  155.42      M  
0         Online        notepad   36.94      M

In [34]:
# Create a pivot table to count the occurrences of 'In-Store' and 'Online' purchase methods by gender
purchase_method_by_gender = pd.crosstab(salesdf['gender'], salesdf['purchaseMethod'])[['In store', 'Online']]

# Remove the index name
purchase_method_by_gender.index.name = "Gender"

print(purchase_method_by_gender)


purchaseMethod  In store  Online
Gender                          
F                   8062    4270
M                   7571    4314


#### Query 5: Monthly total sales

Extract Date First before month

The 'saleDate' column should be converted to datetime objects for proper date manipulation

In [36]:
# Extract the # value of '$date' from the 'saleDate' dictionaries, which are nested within each row.

def extract_date_date(d):
    return pd.to_datetime(d['$date'])

salesdf['sales_date'] = salesdf['saleDate'].apply(extract_date_date)

Displaying Month Sales from 2013 - 2017

In [39]:
# Extract the month from the 'sales_date' column
salesdf['Month'] = salesdf['sales_date'].dt.month

# Extract the year from the 'sales_date' column
salesdf['Year'] = salesdf['sales_date'].dt.year

# Group by Year and Month and calculate the total sales for each month
monthly_sales = salesdf.groupby(['Year', 'Month'])['sales'].sum().reset_index()

# Create a dictionary to map month numbers to full month names
month_names = {
    1: 'January', 2: 'February', 3: 'March', 4: 'April', 5: 'May', 6: 'June',
    7: 'July', 8: 'August', 9: 'September', 10: 'October', 11: 'November', 12: 'December'
}

# Map the Month column to full month names
monthly_sales['Month'] = monthly_sales['Month'].map(month_names)

# Rename the columns for clarity
monthly_sales.columns = ['Year', 'Month', 'Total Sales']

# Print the monthly sales
print(monthly_sales)

    Year      Month  Total Sales
0   2013    January    196043.76
1   2013   February    147622.59
2   2013      March    140700.06
3   2013      April    162336.67
4   2013        May    179143.37
5   2013       June    142132.02
6   2013       July    176298.23
7   2013     August    130977.22
8   2013  September    137871.99
9   2013    October    173126.73
10  2013   November    148358.38
11  2013   December    174306.96
12  2014    January    125824.02
13  2014   February    141174.41
14  2014      March    193413.50
15  2014      April    132593.08
16  2014        May    166896.80
17  2014       June    143304.56
18  2014       July    184804.60
19  2014     August    167979.81
20  2014  September    159802.92
21  2014    October    149284.84
22  2014   November    180564.85
23  2014   December    139466.83
24  2015    January    222577.01
25  2015   February    189776.15
26  2015      March    178086.94
27  2015      April    132298.50
28  2015        May    167482.69
29  2015  

Displaying OVERALL Monthly Sales

In [41]:
# Extract the month from the 'sales_date' column
salesdf['Month'] = salesdf['sales_date'].dt.month

# Create a dictionary to map month numbers to full month names
month_names = {
    1: 'January', 2: 'February', 3: 'March', 4: 'April', 5: 'May', 6: 'June',
    7: 'July', 8: 'August', 9: 'September', 10: 'October', 11: 'November', 12: 'December'
}

# Map the Month column to full month names in the correct order
salesdf['Month'] = pd.Categorical(salesdf['Month'].map(month_names), 
                                  categories=month_names.values(), ordered=True)

# Group by Month and calculate the total sales for each month
monthly_sales = salesdf.groupby('Month')['sales'].sum().reset_index()

# Rename the columns for clarity
monthly_sales.columns = ['Month', 'Total Sales']

# Print the monthly sales
print(monthly_sales)

        Month  Total Sales
0     January    918041.47
1    February    733884.21
2       March    867736.82
3       April    756559.46
4         May    878892.77
5        June    769416.01
6        July    847051.30
7      August    791577.12
8   September    758566.50
9     October    841007.09
10   November    860402.92
11   December    871436.60


### Preperation, Processing & Result of Queries



**Query 1: Top 10 products by sales**

Preperation:

Extracted the 'product_name' from the 'items' nested within each row 

Extracted  'quantity' and 'price' from the 'items' to calculate sales 

Processing: 

Grouped the data by 'product_name.' 

Calculated the total sales (quantity x price) for each product. 

Used nlargest(10) to find the top 10 products by total sales. 


Result: 

Product Name  Total Sales

laptop   - 6775977.07

backpack - 817374.10

pens     - 581843.27

binder   - 511644.57

notepad  - 463615.48

envelopes- 376658.49

printer paper -   367459.29

_____


**Query 2: Top 3 Products by Location**

Preperation: 

Already have the 'product_name' column created in Query 1. 

The 'storeLocation' column available too 

Processing: 

Group the data by 'storeLocation' and 'product_name.' 

Calculate the sum of sales for each product within each store location. 

Used nlargest(3) to find the top 3 products by sales for each store location. 

Result: 

There are 6 locations in total, and when we display 
the top 3 products of each store location, they are the same 3 products: Laptop, Backpack, & Pens
_____


**Query 3: Store location Rankings**

Preperation: 

Have the 'storeLocation' and 'sales' columns 

Processing: 

Group the data by 'storeLocation.' 

Sum the sales for each store. 

Sort the stores by their total sales in descending order. 

Result: 
storeLocation       sales  rank

0        Denver  2921009.92   1.0

1       Seattle  2255947.69   2.0

2        London  1583066.79   3.0

3        Austin  1445603.11   4.0

4      New York  1016059.59   5.0

5     San Diego   672885.17   6.0
_____


**Query 4: Purchased method by Gender table**

 

Preperation: 

New column extract 'gender' from the 'customer' nested within each row. 

Processing:  

Used pd.crosstab() to create a pivot table that counts the occurrences of each purchase method by gender 

Also specified only 2 purchase methods 

Result: 

purchaseMethod  In store  Online                      

F                         8062              4270

M                         7571              4314
_____


**Query 5: Total Monthly sales**

 

Prepeation: 

Extract the 'sales_date' from the 'saleDate' dictionary within each row and convert it to a datetime object. 

Processing: 

Extract the month from the 'sales_date' and map it to full month names                                                                                Group the data by month and calculate the total sales for each month 

Sort the months in chronological order 

Result: 

2 Results -  First Total sales for each month for every year  

Second Overall sales for 12 months 
_____

## Article - Report

The Supplies Sale Data we have been able to ingest, transform & query can all result in very important decision making. If I take the place of this supply store business, and I am able to find out my top products, top location, gender dispargae of sales, when these sales are happening, and how they are happening, I as a business owner can figure out how to expand on this. **Its all about asking the right questions.** 

###### WHO:
This data looks at a a chain stores of supplies list, the type of customer is buying it, where they are buying it, and how they are buying it.
Taking a look at the demographic of people purchasing school/office/workplace supplies can be key in MARKETING. We have found that more female customers purchase instore than male customers. This could tell me as a business owner to change my store displays. Is the physical store not attractive to the male eye? does my online website need a rebrand? are phone calls doing better in sales? Sales data tells a story, and we can use customer demographic results to find out where the business is lacking, and where our budgeting needs to go to improve sales. 


###### WHEN:
We have queried the months in sales, and when we take a look at the overall sales for 2013 til 2017 we may notice a trend. 

January, May, December

These are the top 3 months in sales. One thing to notice is that these 3 months create an even timeline. May is around the middle of a year (summer), and jan and dec are at opposite ends of the year. People are buying these supplies at the start of a working/school year, and they are restocking after summer break. At the end of the year they are replenishing supplies while the year comes to an end. 

###### WHAT:
The categories of supplies fall under: general, school, stationary, office, electronics, kids, writing, travel, and organization. We see this in the table above as a recap.Laptops, bags and pens are the most bought when you consider sales, and this can lead to decision making for the business that fall under stock. A supply business can ask themselves what products are doing the best, which ones are selling the least, and adjust stocking up on product based on what they need more of. 

###### WHERE:
When considering the 6 locations of supply stores, we can also see where the business is selling the most/least.
More decision making can be considered on opening up new stores. Denver does the best in overall sales, while san diego does the least. If this sales pattern resumes the business can consider shutting down the sandiego branch and opening another denver store.


In conclusiona the data analysis for this supply store business tells a lot when it comes to performance. We see which stores are doing the best, we see which gender is purchasing more, we see HOW the gender is purchasing the most, when the most sales are occuring, & last but not least we see which products are sold in the highest quantity. These types of results from queries show we the business that men arent as interested in physical stores. We see that pens bags and laptops are the most popular across all stores. We see that Denver sells the most while sandiego doesnt. We see when people are buying the most, and how they are making these purchases. Analyzing the data of a business in this way showcases and answers a lot of questions, and hopefully it assists this supply business in remaining successful as more years pass.